### PR TITLE
CircuitSimulator Refactor

### PR DESCRIPTION
**Pull request by amccaskey**
Add `applyGate` method for subclasses to implement instead of requiring overloads for all gate methods. 

Introduce gate queue to better support mid-circuit measurement and data persistence to named classical registers. 

Introduce unique cached circuit name to support sub-type caching strategies (e.g. cache the optimal contraction path for a given circuit).

Break up the python quake file check tests into separate ones for easier debugging.

### CODE REVIEW

**Comment by amccaskey** [comment]
I have read the Contributor License Agreement and I hereby accept the Terms.

**Review by boschmitt — COMMENTED** [review]
I like the refactoring, it does simplifies the code. However, it seems that the set of changes is orthogonal; They could have been broken into smaller, more focused PRs (might still be worth doing it).

Breaking up the big python test into smaller ones is a great move, but I would still keep similar ones on the same file, specifically leaving all `qalloc` tests in the same file.

The fix in the clang-format configuration file is likely to affect the whole code base, it certainly worth creating a separate PR that with the fix and rerun in the entire codebase.

File: runtime/nvqir/CircuitSimulator.h [hunk 1]
```diff
@@ -193,26 +396,121 @@ class CircuitSimulator {
...
+    if (sampleQubits.empty())
+      return;
 
-  virtual void setNoiseModel(cudaq::noise_model &noise) {
+    if (executionContext->hasConditionalsOnMeasureResults && !force)
+      return;
+
+    // Sort the qubit indices
+    std::set<std::size_t> tmpSet(sampleQubits.begin(), sampleQubits.end());
+    sampleQubits.assign(tmpSet.begin(), tmpSet.end());
```
[the remark this answers is not in the document]
Comment (amccaskey) [reply runtime/nvqir/CircuitSimulator.h:409]: I was trying to sort and remove potential duplicates. Whenever I think unique elements my head always goes to `set`, and I didn't think about sort / unique. So this is good feedback, I will update with a sort and unique.

File: lib/Target/IQM/TranslateToIQMJson.cpp [hunk 2]
```diff
@@ -6,11 +6,11 @@
  * the terms of the Apache License 2.0 which accompanies this distribution.    *
  *******************************************************************************/
 
+#include "nlohmann/json.hpp"
```
[the remark this answers is not in the document]
Comment (amccaskey) [reply lib/Target/IQM/TranslateToIQMJson.cpp:9]: I will revert the clang-format file change. Should decrease the change set.

File: runtime/nvqir/CircuitSimulator.h [hunk 3]
```diff
@@ -193,26 +396,121 @@ class CircuitSimulator {
...
+          qubitLocMap.insert({qubits[i], idx});
+        }
+
+        cudaq::ExecutionResult tmp(regName);
+        for (auto &[bits, count] : execResult.counts) {
+          std::string b = "";
+          for (auto &qb : qubits)
+            b += bits[qubitLocMap[qb]];
+          tmp.appendResult(b, count);
+        }
```
[the remark this answers is not in the document]
Comment (amccaskey) [reply runtime/nvqir/CircuitSimulator.h:441]: Updated to unordered_map

File: runtime/nvqir/CircuitSimulator.h [hunk 4]
```diff
@@ -8,46 +8,211 @@
...
+#include "Gates.h"
 #include "Logger.h"
 #include "MeasureCounts.h"
 #include "NoiseModel.h"
 #include "QIRTypes.h"
+
 #include <cstdarg>
 #include <cstddef>
+#include <queue>
+#include <set>
```
Comment (boschmitt) [inline runtime/nvqir/CircuitSimulator.h:20]:
```suggestion
```
Suppose we don't need it anymore

**Review by boschmitt — APPROVED** [review]
LGTM, thanks for this work Alex!

File: runtime/nvqir/CircuitSimulator.h (hunk 4 above)
Comment (amccaskey) [reply runtime/nvqir/CircuitSimulator.h:20]: Yea I will remove. I also left some commented code in here, will circle back after I'm done with another branch I'm on :) 

Thanks for the review!

**Pull request merged**